File: python/1-D Dynamic Programming/max_product_subarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # res: stores the maximum product found so far
        res = max(nums)
        # curMin, curMax: track the minimum and maximum product ending at the current position
        curMin, curMax = 1, 1

        for n in nums:
            if n == 0:
                # Reset when encountering zero, as product will be zero
                curMin, curMax = 1, 1
                continue
            temp = curMax * n
            # curMax: max product ending here (could be n, n*curMax, or n*curMin)
            curMax = max(n * curMax, n * curMin, n)
            # curMin: min product ending here (could be n, n*curMax, or n*curMin)
            curMin = min(temp, n * curMin, n)
            # Update result with the best found so far
            res = max(res, curMax, curMin)
        return res

# Time Complexity: O(n), where n is the length of nums. We traverse the list once.
# Space Complexity: O(1), since only a constant amount of extra space is used.

# -------------------
# Test Cases
# -------------------
def test_max_product_subarray():
    s = Solution()
    # Test case 1: Example from Leetcode
    assert s.maxProduct([2,3,-2,4]) == 6  # [2,3]
    # Test case 2: Contains zero
    assert s.maxProduct([-2,0,-1]) == 0
    # Test case 3: All negative numbers
    assert s.maxProduct([-2,-3,-4]) == 12  # [-2,-3,-4]
    # Test case 4: Single element
    assert s.maxProduct([0]) == 0
    assert s.maxProduct([5]) == 5
    # Test case 5: Mix of positive and negative
    assert s.maxProduct([2,-5,-2,-4,3]) == 24  # [-2,-4,3]
    # Test case 6: All zeros
    assert s.maxProduct([0,0,0]) == 0
    print("All test cases passed!")

if __name__ == "__main__":
    test_max_product_subarray()

File: C++/1-D Dynamic Programming/max_product_subarray.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        /*
            Problem:
            - Given an array of integers, find the contiguous subarray with the largest product.

            Approach:
            - Track both current maximum and current minimum products at each position because:
                - A negative number can flip min ↔ max.
            - Reset current products when encountering zero because multiplying further will always give zero.
            - Keep updating the result with the largest product seen so far.

            Time Complexity: O(n)
                - Single pass through the array.
            Space Complexity: O(1)
                - Uses only a few variables.
        */

        int res = *max_element(nums.begin(), nums.end()); 
        // `res` stores the maximum product found so far (handles case where all numbers are ≤ 0)

        int curMin = 1, curMax = 1; 
        // Track min and max products ending at the current index

        for (int n : nums) {
            if (n == 0) {
                // If current element is zero, reset curMin & curMax to 1
                curMin = 1;
                curMax = 1;
                continue;
            }

            int temp = curMax * n;
            // Save curMax * n before updating curMax

            curMax = max({n * curMax, n * curMin, n});
            // Maximum product ending here can be: n, n * curMax (prev), or n * curMin (prev)

            curMin = min({temp, n * curMin, n});
            // Minimum product ending here can also help if multiplied with another negative later

            res = max({res, curMax, curMin});
            // Update the result with the largest seen so far
        }

        return res;
    }
};

/*
================ LINE-BY-LINE SYNTAX EXPLANATION =================

#include <bits/stdc++.h>
    - Includes all standard C++ headers (like vector, iostream, etc).

using namespace std;
    - Allows us to use standard names (like vector, string) without prefixing them with std::.

class Solution {
    - Defines a class named Solution.

public:
    - The methods below are publicly accessible.

int maxProduct(vector<int>& nums)
    - Defines a function `maxProduct` that takes a vector of integers `nums` and returns an int.

int res = *max_element(nums.begin(), nums.end());
    - Initializes `res` with the largest number in `nums` to handle arrays where all numbers are ≤ 0.

int curMin = 1, curMax = 1;
    - Variables to keep track of the current minimum and maximum product ending at this index.

for (int n : nums)
    - Loop over each integer `n` in the vector `nums`.

if (n == 0)
    - If `n` is zero, reset `curMin` and `curMax` because zero resets the product.

int temp = curMax * n;
    - Temporarily store `curMax * n` before it gets updated (used later for curMin).

curMax = max({n * curMax, n * curMin, n});
    - Update `curMax` with the maximum of the three options.

curMin = min({temp, n * curMin, n});
    - Update `curMin` with the minimum of the three options.

res = max({res, curMax, curMin});
    - Update the result `res` if we find a bigger product.

return res;
    - Return the largest product found.

*/
